examples/sugarscape-ig/ss_pandas/agents.py

```diff
@@ -0,0 +1,129 @@
+import numpy as np
+import pandas as pd
+from mesa_frames import AgentSetPandas, ModelDF
+
+
+class AntPandas(AgentSetPandas):
+    def __init__(
+        self,
+        model: ModelDF,
+        n_agents: int,
+        initial_sugar: np.ndarray | None = None,
+        metabolism: np.ndarray | None = None,
+        vision: np.ndarray | None = None,
+    ):
+        super().__init__(model)
+
+        if initial_sugar is None:
+            initial_sugar = model.random.integers(6, 25, n_agents)
+        if metabolism is None:
+            metabolism = model.random.integers(2, 4, n_agents)
+        if vision is None:
+            vision = model.random.integers(1, 6, n_agents)
+
+        agents = pd.DataFrame(
+            {
+                "unique_id": np.arange(n_agents),
+                "sugar": model.random.integers(6, 25, n_agents),
+                "metabolism": model.random.integers(2, 4, n_agents),
+                "vision": model.random.integers(1, 6, n_agents),
+            }
+        )
+        self.add(agents)
+
+    def move(self):
+        neighborhood: pd.DataFrame = self.space.get_neighborhood(
+            radius=self["vision"], agents=self, include_center=True
+        )
+
+        # Merge self.space.cells to obtain properties ('sugar') per cell
+        neighborhood = neighborhood.merge(self.space.cells, on=["dim_0", "dim_1"])
+
+        # Merge self.pos to obtain the agent_id of the center cell
+        # TODO: get_neighborhood/get_neighbors should return 'agent_id_center' instead of center position when input is AgentLike
+        neighborhood["agent_id_center"] = neighborhood.merge(
+            self.pos.reset_index(),
+            left_on=["dim_0_center", "dim_1_center"],
+            right_on=["dim_0", "dim_1"],
+        )["unique_id"]
+
+        # Order of agents moves based on the original order of agents.
+        # The agent in his cell has order 0 (highest)
+        agent_order = neighborhood.groupby(["agent_id_center"], sort=False).ngroup()
+        neighborhood["agent_order"] = agent_order
+        agent_order = neighborhood[["agent_id_center", "agent_order"]].drop_duplicates()
+
+        neighborhood = neighborhood.merge(
+            agent_order.rename(
+                columns={
+                    "agent_id_center": "agent_id",
+                    "agent_order": "blocking_agent_order",
+                }
+            ),
+            on="agent_id",
+        )
+
+        # Filter impossible moves
+        neighborhood = neighborhood[
+            neighborhood["agent_order"] >= neighborhood["blocking_agent_order"]
+        ]
+
+        # Sort cells by sugar and radius (nearest first)
+        neighborhood = neighborhood.sort_values(
+            "sugar", "radius", ascending=[False, True]
+        )
+
+        best_moves = pd.DataFrame()
+
+        # While there are agents that do not have a best move, keep looking for one
+        while len(best_moves) < len(self.agents):
```

Comment:
This is awesome, a hybrid of sequential and GPU/CPU parallelizable executions (whenever there is no collision). We should use this pattern a lot more in other models. I would want to know how many iterations this does across the steps, as a function of (num_agents / (width * height)).

I can't remember how the FLAME 2 folks did it, or how the Sugarscape on steroid folks did it. I think the former did tie-breaking whenever there is a collision?

Reply:
> I would want to know how many iterations this does across the steps, as a function of (num_agents / (width * height)).

This is optional, by the way. There are a lot more other pressing stuff.

Reply:
> This is awesome, a hybrid of sequential and GPU/CPU parallelizable executions (whenever there is no collision). We should use this pattern a lot more in other models. I would want to know how many iterations this does across the steps, as a function of (num_agents / (width * height)).
> 
> I can't remember how the FLAME 2 folks did it, or how the Sugarscape on steroid folks did it. I think the former did tie-breaking whenever there is a collision?

Yes, you mentioned that, I don't know about enough CUDA to look through the code :).
When prototyping with 100 agents 10*10 space (so full of collisions) was about 4 iterations.

Reply:
That's really good to know. I'm going over the CUDA code right now. I approved this PR so that we can move faster. Any direct comparison with the same algorithm as FLAME 2 can happen later.

Reply:
The movement diagram for FLAME GPU 2 can be found at figure 6 of [the paper](https://onlinelibrary.wiley.com/doi/10.1002/spe.3207):
1. [Movement request](https://github.com/FLAMEGPU/FLAMEGPU2-submodel-benchmark/blob/290702073492677f2fc9164419e9d5900d119e02/src/main.cu#L101-L144)
2. [Movement response](https://github.com/FLAMEGPU/FLAMEGPU2-submodel-benchmark/blob/290702073492677f2fc9164419e9d5900d119e02/src/main.cu#L145-L187), if there is a conflict, tie breaking happens, where the priority is chosen at random `float message_priority = FLAMEGPU->random.uniform<float>()`.
3. [Movement transaction](https://github.com/FLAMEGPU/FLAMEGPU2-submodel-benchmark/blob/290702073492677f2fc9164419e9d5900d119e02/src/main.cu#L188-L216)

Based on what they said:
> The cell responds by selecting the agent with highest priority therefore allowing a single agent to move in the first iteration of the model. An exit condition only allows the main simulation to resume once all agents are ‘resolved’ and as such a second iteration of the sub-model is required for the second agent to find a collision free location (and the exit condition to pass).

That "2nd iteration" could probably happen several times as long as there are still conflicts, I think? Their algorithm might be very similar to yours.